python/nav/web/auth/ldap.py

```diff
@@ -121,11 +122,66 @@ def open_ldap():
     return lconn
 
 
-def authenticate(login, password):
+def authenticate(username, password):
     """
-    Attempt to authenticate the login name with password against the
-    configured LDAP server.  If the user is authenticated, required
-    group memberships are also verified.
+    Authenticate the username and password against the configured LDAP server.
+
+    Required group memberships are also verified.
+
+    Returns an authenticated Account with updated groups, or None.
+    """
+    if not available:
+        return None
+    ldap_user = get_ldap_user(username, password)
+    try:
+        account = Account.objects.get(login__iexact=username, ext_sync='ldap')
+    except Account.DoesNotExist:
+        if ldap_user:
+            account = autocreate_ldap_user(ldap_user, password)
+            return account
+    if account.locked:
+        _logger.info("Locked user %s tried to log in", account.login)
+        return None
+    if account.check_password(password):
+        account = update_ldap_user(ldap_user, account, password)
+        return account
```

Comment:
This got confusing, real fast.

I read this as: If a re-visiting user is authenticated via LDAP, this additionally checks their entered password against the password stored in the DB.  If their LDAP password changed at some point, this local verification will fail, and the user can no longer log in to NAV.

Then I realized that this function is actually only ever called for first-time users.  The function contains code to handle the difference between new users and re-visiting users.  It won't work well for re-visiting users, if I'm reading it correctly, but it will also never be called for re-visiting users, so what's the plan here?

Reply:
It is confusing: it's partially duplicating what `nav.web.auth.authenticate` as a small step to make the latter independent of ldap.

It will be called for re-visiting users once we split up `nav.web.auth.authenticate`. Then it will also change: if ldap is on:

* If ldap_user is False,  return None and fallback to local user
* If ldap_user is not False it *is* authenticated, get the account and return the account

See my comment above about the fact that what happens if `ldap_user is False` currently isn't covered and that there are multiple ways to do it.

Reply:
So, is the change you want a flow-chart?